cmd/kubeadm/app/discovery/token/token.go

```diff
@@ -179,20 +182,27 @@ func buildSecureBootstrapKubeConfig(endpoint string, caCert []byte) *clientcmdap
 }
 
 // runForEndpointsAndReturnFirst loops the endpoints slice and let's the endpoints race for connecting to the master
-func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(string) (*clientcmdapi.Config, error)) *clientcmdapi.Config {
+func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(string) (*clientcmdapi.Config, error)) (*clientcmdapi.Config, error) {
 	stopChan := make(chan struct{})
 	var resultingKubeConfig *clientcmdapi.Config
 	var once sync.Once
 	var wg sync.WaitGroup
+	var resultingError error
 	for _, endpoint := range endpoints {
 		wg.Add(1)
 		go func(apiEndpoint string) {
+			retries := constants.DiscoveryTimeout / constants.DiscoveryRetryInterval
 			defer wg.Done()
 			wait.Until(func() {
 				fmt.Printf("[discovery] Trying to connect to API Server %q\n", apiEndpoint)
 				cfg, err := fetchKubeConfigFunc(apiEndpoint)
 				if err != nil {
 					fmt.Printf("[discovery] Failed to connect to API Server %q: %v\n", apiEndpoint, err)
+					retries--
+					if retries == 0 {
+						resultingError = err
```

Comment:
Better to add a customized message to the error, like `[discovery] Aborted after retrying xx times: <error msg here>`. WDYT?